89. Gray Code

```diff
@@ -0,0 +1,26 @@
+//Time: O (2 ^ n) Space: O(2 ^ n)
```

Comment:
Give me a few test cases, and show me the result that you think your code will output. The test case should cover all scenarios as much as possible.

Reply:
If you find that your result is not what you want, then it is time to go through your algorithm again. Proposal test cases and run it through is an essential step in an interview.

Reply:
1. It is guaranteed that n is non-negative.
2. n = 0 will return an empty vector.
3. n = 1 will return [0, 1]
4. Let's say n = 4, it is expected that we will have 16 numbers.
   First eight numbers are result of n = 3, [0000, 0001, 0011,  0010, 0110, 0111, 0101, 0100]
   And the second half will be the reverse of the first half with an addition 1 in the front
  [1100, 1101, 1111, 1110, 1010, 1011, 1001, 1000]
5. For the n = 32 case, we are only handling the bits, so I am not worried about overflow or whatever.
5. The function return type is pre-defined (vector<int>), which means we cannot handle the case n > 32. I dont know how to deal with this so just return an empty vector.